cloudfoundry/resource_cf_service_instance.go

```diff
@@ -136,6 +136,13 @@ func resourceServiceInstanceUpdate(d *schema.ResourceData, meta interface{}) (er
 
 	session.Log.DebugMessage("begin resourceServiceInstanceUpdate")
 
+	// Enable partial state mode
+	// We need to explicitly set state updates ourselves or
+	// tell terraform when a state change is applied and thus okay to persist
+	// In particular this is necessary for params since we cannot query CF for
+	// the current value of this field
+	d.Partial(true)
```

Comment:
https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceData.Partial mentions 
> When partial state mode is enabled, then only key prefixes specified by SetPartial will be in the final state. This allows providers to return partial states for partially applied resources (when errors occur). 

I'd asssume next steps for service instance partial mode will be to call `setPartial` with service related fields, otherwise no fields will be saved.

However, since in this case the `resourceServiceInstanceUpdate` function contains a single `sm.UpdateServiceInstance` api call, I'm not sure we need to enable partial mode: Terraform will save the full state as mentionned in the [doc](https://www.terraform.io/docs/extend/writing-custom-providers.html) upon function return 

> If the Update callback returns with or without an error, the full state is saved

Reply:
Given that we do all of the updates as a single API call to Cloud Foundry, we don't need to flag individual elements.  The call `d.Partial(false)` later on, returns terraform to the current behaviour of propagating all expected changes to the state file in the case of success.  

In the case that Cloud Foundry actually only performed a partial update, we'd realize this during the next refresh phase for all properties except for the `json_params`, which we cannot read out, and terraform would then realise that it needed to reattempt the update.  In the case that the `json_params` were actually updated while another change was made, we'll repush this change, but that shouldn't do any harm.

Reply:
The bottom line is that, this really only becomes an issue for properties that we cannot read out from the API during the refresh phase, like `json_params`.  In these cases we need to be careful to ensure that the state file correctly reflects the current state of the real system, to the best of our ability.

Reply:
ok thanks Jim for the clarification. Looks good to me.